### PR TITLE
fix(event): updated event.Data

### LINKED ISSUE #1327: Remove unnecesary Event.Data encoding/decoding
Change the [Event.Data](https://github.com/0chain/0chain/blob/2d587247e7168f494277f73970c2c1694b010ab5/code/go/0chain.net/smartcontract/dbs/event/event.go#L18) type from `[]byte` to `interface{}` and do type assertion to get concrete event data in [addStat](https://github.com/0chain/0chain/blob/2d587247e7168f494277f73970c2c1694b010ab5/code/go/0chain.net/smartcontract/dbs/event/process.go#L90). We don't need to do the JSON encoding/decoding work.

### LINKED ISSUE #1451: sql error on allocation_fetch
```

./zbox upload --localpath ~/Downloads/ImageOptim.tbz2 --allocation $alloc --remotepath /ImageOptim1.tbz2                                                                                                     11:09:38
0chain-core-sdk [INFO]   2022/06/29 11:16:48.876529 wallet.go:292: ******* Wallet SDK Version:v1.7.2-2-g3237f5c-dirty ******* (SetLogFile)
0box-sdk        [INFO]   2022/06/29 11:16:48.879494 sdk.go:73: ******* Storage SDK Version: v1.7.2-2-g3237f5c-dirty *******
0chain-core-sdk [INFO]   2022/06/29 11:16:51.159487 wallet.go:415: ******* Wallet SDK Version:v1.7.2-2-g3237f5c-dirty ******* (InitZCNSDK)
0box-sdk        [DEBUG]  2022/06/29 11:16:52.767852 networkworker.go:105: Get network result:{"miners":["https://test.0chain.net/miner03","https://test.0chain.net/miner02","https://test.0chain.net/miner01"],"sharders":["https://test.0chain.net/sharder01","https://test.0chain.net/sharder02"]}
Error fetching the allocation. allocation_fetch_error: Error fetching the allocation.internal_error: can't get allocation: error retrieving allocation: 8a274a48f4ff78adb24b8fa9a074832926ea33d20f941f37772655d46ea9fbda, error: sql: Scan error on column index 16, name "read_price_max": converting driver.Value type int64 ("-9223372036854775808") to a uint64: invalid syntax
```
detail https://0chain.slack.com/archives/C01GMABG2T1/p1656472583372719?thread_ts=1656464433.865429&cid=C01GMABG2T1

### LINKED ISSUE #1452: Fix number inconsistencies due to conversion from uint64 to float64 to int64 and back to uint64.
When `encoding/json` unmarshalls into `map[string]interface{}` it type casts all the numbers to float64. Type casting to float64 will cause rounding errors i.e. it rounds off to close number that it can represent with float64. https://go.dev/play/p/wpZgX4fOU59

In https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/dbs/types.go#L7, the type of `Updates` field is `map[string]interface{}` so when unmarshalling fields such as `Balance` it will type cast to float64.

Error in system-test:
```
./zbox upload --localpath ~/Downloads/ImageOptim.tbz2 --allocation $alloc --remotepath /ImageOptim1.tbz2                                                                                                     11:09:38
0chain-core-sdk [INFO]   2022/06/29 11:16:48.876529 wallet.go:292: ******* Wallet SDK Version:v1.7.2-2-g3237f5c-dirty ******* (SetLogFile)
0box-sdk        [INFO]   2022/06/29 11:16:48.879494 sdk.go:73: ******* Storage SDK Version: v1.7.2-2-g3237f5c-dirty *******
0chain-core-sdk [INFO]   2022/06/29 11:16:51.159487 wallet.go:415: ******* Wallet SDK Version:v1.7.2-2-g3237f5c-dirty ******* (InitZCNSDK)
0box-sdk        [DEBUG]  2022/06/29 11:16:52.767852 networkworker.go:105: Get network result:{"miners":["https://test.0chain.net/miner03","https://test.0chain.net/miner02","https://test.0chain.net/miner01"],"sharders":["https://test.0chain.net/sharder01","https://test.0chain.net/sharder02"]}
Error fetching the allocation. allocation_fetch_error: Error fetching the allocation.internal_error: can't get allocation: error retrieving allocation: 8a274a48f4ff78adb24b8fa9a074832926ea33d20f941f37772655d46ea9fbda, error: sql: Scan error on column index 16, name "read_price_max": converting driver.Value type int64 ("-9223372036854775808") to a uint64: invalid syntax

```

For tests with gorm please check https://go.dev/play/p/RVxnninRu1p.

### PR DESCRIPTION
**Pull request by cnlangzi**
## Fixes
- fixed https://github.com/0chain/0chain/issues/1327
## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- .

### CODE REVIEW

File: code/go/0chain.net/smartcontract/dbs/event/process.go [hunk 1]
```diff
@@ -73,10 +75,20 @@ func (edb *EventDb) addEventsWorker(ctx context.Context) {
 			case TypeStats:
 				err = edb.addStat(event)
 			case TypeError:
-				err = edb.addError(Error{
-					TransactionID: event.TxHash,
-					Error:         event.Data,
-				})
+
+				msg, ok := fromEvent[string](event.Data)
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/dbs/event/process.go]: We don't need to use generics to get the data, `fmt.Sprintf("%v", event.Data)` is simpler and should have  the same effort.

File: code/go/0chain.net/smartcontract/stakepool/edb_delegate_pool.go [hunk 2]
```diff
@@ -34,31 +32,33 @@ func (dp DelegatePool) emitNew(
...
 		event.TypeStats,
 		event.TagUpdateDelegatePool,
 		dpu.PoolId,
-		string(data),
+		delegatePoolUpdateToDbsDelegatePoolUpdate(dpu),
 	)
 	return nil
 }
+
+func delegatePoolUpdateToDbsDelegatePoolUpdate(dpu DelegatePoolUpdate) dbs.DelegatePoolUpdate {
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/stakepool/edb_delegate_pool.go:59]: This function name is too long and repeating, perhaps use `newDBDelegatePoolUpdate()`

File: code/go/0chain.net/smartcontract/stakepool/edb_stakepool.go [hunk 3]
```diff
@@ -25,15 +23,20 @@ func (spu StakePoolReward) Emit(
...
 		event.TypeStats,
 		tag,
 		spu.ProviderId,
-		string(data),
+		stakePoolRewardToStakePoolRewardEvent(spu),
 	)
 	return nil
 }
+
+func stakePoolRewardToStakePoolRewardEvent(spu StakePoolReward) *dbs.StakePoolReward {
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/stakepool/edb_stakepool.go:36]: newStakePoolRewardEvent()

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/dbs/event/process.go (hunk 1 above)
Comment (cnlangzi) [reply code/go/0chain.net/smartcontract/dbs/event/process.go]: fixed

File: code/go/0chain.net/smartcontract/stakepool/edb_delegate_pool.go (hunk 2 above)
Comment (cnlangzi) [reply code/go/0chain.net/smartcontract/stakepool/edb_delegate_pool.go:59]: fixed

File: code/go/0chain.net/smartcontract/stakepool/edb_stakepool.go (hunk 3 above)
Comment (cnlangzi) [reply code/go/0chain.net/smartcontract/stakepool/edb_stakepool.go:36]: fixed

**Review by peterlimg — APPROVED** [review]
lgtm, please fix the conflicts

**Comment by peterlimg** [comment]
please also update the new added event for adding users here https://github.com/0chain/0chain/pull/1234/files#diff-c53991643894d22c975d85372445b129e0882fd55d8fa6229a82fe07b4d3c80fR708-R718, thanks.

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | fix/delegate_pool_issue  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Pull request merged**